### PR TITLE
Updated time to travel distance helper functions

### PR DESCRIPTION
**Pull request by nimazareian**
### Description
Updated the implementation of time to travel distance to incorporate initial and final velocities. 
This can be used in the new primitive assignment architecture to calculate a more realistic tactic cost than just path distance (e.g. if a robot is closer to tactic destination but is already driving away from it should have a higher cost than a robot that is slightly further away but is driving towards the destination). The previous implementation was not viable for the primitive assignment algorithm as the time to destination was directly correlated to the distance. 
This function was already being used in intercepting passes and dribble FSM, and some cost function. Currently the usages there are not changed (i.e. still uses 0 initial and final velocity), but they can be updated to incorporate the updated features. 

Renamed the file which the general calculation live in to `time_to_travel` from `pass`. Added helper wrapper functions to `Robot` class to make it easier to use.

### Testing Done
Added new tests and previously added tests are still passing

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/evaluation/time_to_travel.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include "software/time/duration.h"
+
+/**
+ * Calculate minimum time it would take for an robot to move a set distance
```
Comment (amrm3lm) [inline src/software/ai/evaluation/time_to_travel.h]: a robot

File: src/software/ai/evaluation/time_to_travel.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * The calculations are done in 1D to simplify calculations and be as performant as
+ * possible
+ *
+ * @param distance The distance which the robot is moving
+ * @param max_velocity The maximum linear velocity the robot can travel at (m/s)
```
Comment (amrm3lm) [inline src/software/ai/evaluation/time_to_travel.h]: i would reword this to make it more general, since youve used this method with a rad/s input in `robot::getTimeToOrientation`

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                                 const double max_acceleration,
+                                 const double initial_velocity,
+                                 const double final_velocity)
+{
+    // Bound all values to be realistic
+    double d_total = std::max(0.0, distance);
+    double v_max   = std::max(0.0, max_velocity);
+    double v_i     = std::clamp(initial_velocity, -max_velocity, max_velocity);
+    double v_f     = std::clamp(final_velocity, 0.0, max_velocity);
+    double a_max   = std::max(1e-6, max_acceleration);
```
Comment (amrm3lm) [inline src/software/ai/evaluation/time_to_travel.cpp:16]: whats the reason for using `1e-6`?

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    // Minimum distance required to accelerate/decelerate from initial to final velocity
+    double dist_required_to_reach_v_f =
+        std::abs(std::pow(v_f, 2) - std::pow(v_i, 2)) / (2 * a_max);
+    if (dist_required_to_reach_v_f > d_total)
```
Comment (amrm3lm) [inline src/software/ai/evaluation/time_to_travel.cpp:21]: im not sure what happens here when `a_max` is `1e-6`, wouldnt we be trying to change velocities without accelerating?

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        // Accelerating/decelerating instantly from initial to final velocity is not
+        // possible with in the given distance, given the robot's max acceleration,
```
Comment (amrm3lm) [inline src/software/ai/evaluation/time_to_travel.cpp]: within

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        double t_total =
+            (-v_i + std::sqrt(std::pow(v_i, 2) + 2 * a_max_signed * d_total)) /
+            a_max_signed;
+        return Duration::fromSeconds(t_total);
+    }
+
+    // Total travel time if the robot is always accelerating at max acceleration (i.e.
+    // accelerating to the highest value possible given the distance, and then
+    // decelerating to final velocity)
+    double t_total =
```
Comment (amrm3lm) [inline src/software/ai/evaluation/time_to_travel.cpp:49]: curious what the derivation here is. i thought kinematic equations only applied when there is constant acceleration, but my understanding from the comment above is that you are accelerating and then decelerating?

**Review by amrm3lm — COMMENTED** [review]
lgtm just wanted some clarification on a couple things

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+Duration getTimeToTravelDistance(const double distance, const double max_velocity,
+                                 const double max_acceleration,
+                                 const double initial_velocity,
+                                 const double final_velocity)
+{
+    // Bound all values to be realistic
+    double d_total = std::max(0.0, distance);
+    double v_max   = std::max(0.0, max_velocity);
+    double v_i     = std::clamp(initial_velocity, -max_velocity, max_velocity);
+    double v_f     = std::clamp(final_velocity, 0.0, max_velocity);
```
Comment (itsarune) [inline src/software/ai/evaluation/time_to_travel.cpp:15]: why is initial velocity clamped to +/- initial velocity but final velocity clamped to [0, max_velocity]

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        if (v_f < v_i)
+        {
+            // Robot must decelerate, so it has negative acceleration
+            a_max_signed *= -1.0;
+        }
+        // t = (-Vi + sqrt(Vi^2 + 2 * a * d)) / a
+        double t_total =
+            (-v_i + std::sqrt(std::pow(v_i, 2) + 2 * a_max_signed * d_total)) /
+            a_max_signed;
+        return Duration::fromSeconds(t_total);
```
Comment (itsarune) [inline src/software/ai/evaluation/time_to_travel.cpp:40]: could you double check this for the case where v_f < v_i? I'm worried that we should be looking at the solution where it's (-v_i - std::sqrt...) if the acceleration is signed as negative

File: src/software/ai/evaluation/time_to_travel_test.cpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+};
+
+TEST_F(TimeToTravel, getTimeToTravelDistance_already_at_dest)
+{
+    EXPECT_EQ(Duration::fromSeconds(0), getTimeToTravelDistance(0, 2.0, 3.0));
+}
+
+TEST_F(TimeToTravel, getTimeToTravelDistance_reaches_max_velocity)
+{
+    // Check that the robot reaches the dest in the at the expected time when
```
Comment (itsarune) [inline src/software/ai/evaluation/time_to_travel_test.cpp]: nit: "reaches the dest at the expected time"

**Review by itsarune — COMMENTED** [review]
great work, but I'm unsure of some of the kinematics when it comes to the final velocity being  negative (it doesn't look like we support it right now)

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 7 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp:15]: Initial velocity can be negative since initially you may be driving away from your destination, however, final velocity didn't make much sense to me when its negative. I think your velocity would be either 0 or positive to pass through the destination, if it's negative (the robot is coming back), I would imagine that you would calculate the time to travel the _path_ in two steps:
1) going to the first destination and stopping at it (Vf = 0) 
2) Starting with Vi = 0 at first destination and going to the second destination point which would bring the robot back.

That was the assumption I went off of.

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 3 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp:16]: There were two things, first being that _max_ acceleration possible by the robot should be a non-negative number, second being that we're doing lots of divisions by acceleration in the formulas, so I did not want the max acceleration to be 0 to avoid SIGFPEs.

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 4 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp:21]: Well we would technically be accelerating, but very slowly. So the distance required to accelerate to Vf would be very large (i.e. `dist_required_to_reach_v_f`). Which would also make sense as we're diving by a very small positive number (i.e. `1e-6`)

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 6 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp:49]: Yeah you're correct. I wasn't sure how exactly I should go about explaining this. 
The formula is derived from combining regular 1D kinematics equations under the assumptions that the robot will travel a distance as fast as possible (i.e. Accelerating as much as possible, and decelerating as late as possible to reach the desired final velocity). Under those assumptions I derived the formula for the total time. 
To help visualize it, Ive made this [desmos](https://www.desmos.com/calculator/jz9txhttq9) link with a velocity vs time graph, where the area under the graph should be equal to the distance we want to travel.

Comment (amrm3lm) [reply src/software/ai/evaluation/time_to_travel.cpp:49]: this was very helpful. i was also afraid the equations wouldnt behave well near `a=1e-6` but i can see now that the limits work out

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 7 above)
Comment (amrm3lm) [reply src/software/ai/evaluation/time_to_travel.cpp:15]: i think the condition `vf>=0` makes sense. To make it explicit maybe we should mention it in the method header.

Comment (itsarune) [reply src/software/ai/evaluation/time_to_travel.cpp:15]: that makes sense to me now. The velocity is relative to the robot; in my mind I was imagining the that a robot travelling in the -X direction would have a negative velocity (aka velocity relative to the field)

File: src/software/ai/evaluation/time_to_travel.h (hunk 1 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.h]: Updated

File: src/software/ai/evaluation/time_to_travel.h (hunk 2 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.h]: Updated the comments to use general units

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 5 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp]: updated

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 7 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp:15]: Added the constraint to the header

File: src/software/ai/evaluation/time_to_travel_test.cpp (hunk 9 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel_test.cpp]: Updated

File: src/software/world/robot.cpp [hunk 10]
```diff
@@ -170,3 +171,31 @@ Polygon Robot::dribblerArea() const
              vector_to_front.perpendicular().normalize(width),
          bottom_left_position + vector_to_front.perpendicular().normalize(width)});
 }
+
+
+Duration Robot::getTimeToOrientation(const Angle &desired_orientation) const
```
Comment (jonathanlew) [inline src/software/world/robot.cpp]: what about final angular velocity?

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+#include <cmath>
+
+Duration getTimeToTravelDistance(const double distance, const double max_velocity,
+                                 const double max_acceleration,
+                                 const double initial_velocity,
+                                 const double final_velocity)
+{
+    // Bound all values to be realistic
+    double d_total = std::max(0.0, distance);
+    double v_max   = std::max(0.0, max_velocity);
```
Comment (jonathanlew) [inline src/software/ai/evaluation/time_to_travel.cpp:13]: this rejection of negative should be part of the function comments

File: src/software/world/robot.cpp [hunk 12]
```diff
@@ -170,3 +171,31 @@ Polygon Robot::dribblerArea() const
...
+{
+    double dist = orientation().minDiff(desired_orientation).toRadians();
```
Comment (jonathanlew) [inline src/software/world/robot.cpp:179]: so we're implicitly finding positive values? Can we have a test where desired orientation is less than the current orientation?

**Review by jonathanlew — COMMENTED** [review]
pretty clean

File: src/software/world/robot.cpp (hunk 12 above)
Comment (nimazareian) [reply src/software/world/robot.cpp:179]: Added a new test where the destination angle is less than the starting angle

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 11 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp:13]: Added it to the javadoc

File: src/software/world/robot.cpp (hunk 10 above)
Comment (nimazareian) [reply src/software/world/robot.cpp]: I didn't really see when you would use it, but I guess there might be a scenario that I can't forsee, so I added it as a optional parameter.

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 8 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp:40]: I looked abit more into it and I think just taking the positive case is okay.

File: src/software/ai/evaluation/time_to_travel_test.cpp [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        std::abs(std::pow(final_vel, 2) - std::pow(initial_vel, 2)) / (2 * max_accel);
+    double distance = dist_required_to_reach_v_f - 0.00001;
+    std::cout << "dist=" << distance << std::endl;
+
+    // The robot is decelerating throughout the path (since the final velocity is too
+    // small for it to reach within the distance) t = (-Vi + sqrt(Vi^2 + 2 * a * d)) / a,
+    // with a being negative due to deceleration
+    //    double acceleration_time =
+    //        (-initial_vel + std::sqrt(std::pow(initial_vel, 2) - 2 * max_accel *
+    //        distance)) / -max_accel;
```
Comment (itsarune) [inline src/software/ai/evaluation/time_to_travel_test.cpp]: you're good to remove these lines from the code since you added the formula in comments above this anyway

**Review by itsarune — COMMENTED** [review]
looks good, one nit from me

File: src/software/ai/evaluation/time_to_travel_test.cpp (hunk 13 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel_test.cpp]: Good catch. I've removed it

File: src/software/ai/evaluation/time_to_travel_test.cpp [hunk 14]
```diff
@@ -186,20 +162,20 @@ TEST_F(
     double final_vel   = 0.0;
     double max_accel   = 1.0;
     double max_vel     = 4.0;
-    double distance    = 1.0;
+    double dist_required_to_reach_v_f =
+        std::abs(std::pow(final_vel, 2) - std::pow(initial_vel, 2)) / (2 * max_accel);
+    double distance = dist_required_to_reach_v_f - 0.00001;
+    std::cout << "dist=" << distance << std::endl;
```
Comment (jonathanlew) [inline src/software/ai/evaluation/time_to_travel_test.cpp]: debugging print out should be removed before merging

File: src/software/ai/evaluation/time_to_travel.cpp [hunk 15]
```diff
@@ -30,10 +31,13 @@ Duration getTimeToTravelDistance(const double distance, const double max_velocit
...
             a_max_signed *= -1.0;
         }
-        // t = (-Vi + sqrt(Vi^2 + 2 * a * d)) / a
+        // Following formula is solved by solving for t in:
+        // d = Vi*t + 1/2*at^2  ───── solve for t ─────┐
+        // t = (-Vi + sqrt(Vi^2 + 2 * a * d)) / a  ◄───┘
         double t_total =
             (-v_i + std::sqrt(std::pow(v_i, 2) + 2 * a_max_signed * d_total)) /
             a_max_signed;
+        std::cout << t_total << std::endl;
```
Comment (jonathanlew) [inline src/software/ai/evaluation/time_to_travel.cpp]: debugging print out should be removed before merging

File: src/software/ai/evaluation/time_to_travel_test.cpp (hunk 14 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel_test.cpp]: good catch. removed

File: src/software/ai/evaluation/time_to_travel.cpp (hunk 15 above)
Comment (nimazareian) [reply src/software/ai/evaluation/time_to_travel.cpp]: removed

**Review by amrm3lm — COMMENTED** [review]
lgtm pending checks

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Review by amrm3lm — APPROVED** [review]
LGTM good work!
